entrypoint/exec/go_init.go

```diff
@@ -0,0 +1,124 @@
+
+MIT License
+
+Copyright (c) 2017 Pablo RUTH
+Copyright (c) 2020 Red Hat, Inc
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+package exec
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+func RemoveZombies(ctx context.Context, wg *sync.WaitGroup) {
+	for {
+		var status syscall.WaitStatus
+
+		// Wait for orphaned zombie process
+		pid, _ := syscall.Wait4(-1, &status, syscall.WNOHANG, nil)
+
+		if pid <= 0 {
+			// PID is 0 or -1 if no child waiting
+			// so we wait for 1 second for next check
+			time.Sleep(1 * time.Second)
+		} else {
+			// PID is > 0 if a child was reaped
+			// we immediately check if another one
+			// is waiting
+			continue
+		}
+
+		// Non-blocking test
+		// if context is done
+		select {
+		case <-ctx.Done():
+			// Context is done
+			// so we stop goroutine
+			wg.Done()
+			return
+		default:
+		}
+	}
+}
+
+func Run(cmd *exec.Cmd) error {
+	// Register chan to receive system signals
+	sigs := make(chan os.Signal, 1)
+	defer close(sigs)
+	signal.Notify(sigs)
+	defer signal.Reset()
+
+	// Define command and rebind
+	// stdout and stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	// Create a dedicated pidgroup
+	// used to forward signals to
+	// main process and all children
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+
+	// Handler for ctrl-c
+	c := make(chan os.Signal)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		syscall.Kill(-cmd.Process.Pid, syscall.SIGTERM)
+		os.Exit(1)
+	}()
+
+	// Goroutine for signals forwarding
+	go func() {
+		for sig := range sigs {
+			// Ignore SIGCHLD signals since
+			// thez are only usefull for go-init
+			if sig != syscall.SIGCHLD {
+				// Forward signal to main process and all children
+				syscall.Kill(-cmd.Process.Pid, sig.(syscall.Signal))
+			}
+		}
+	}()
+
+	// Start defined command
+	err := cmd.Start()
+	if err != nil {
+		return err
+	}
+
+	// Wait for command to exit
+	err = cmd.Wait()
```

Comment:
Doing this "start + wait + exit" pattern I think should always raise the question of why we're not just directly `exec()`-ing the child process.

Reply:
`run-steps` can run N files; we wait to see if that the command is done.